circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/RetainedStateHolder.kt

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2024 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.foundation
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.key
+import androidx.compose.runtime.remember
+import com.slack.circuit.foundation.internal.withCompositionLocalProvider
+import com.slack.circuit.retained.CanRetainChecker
+import com.slack.circuit.retained.LocalCanRetainChecker
+import com.slack.circuit.retained.LocalRetainedStateRegistry
+import com.slack.circuit.retained.RetainedStateRegistry
+import com.slack.circuit.retained.RetainedValueProvider
+import com.slack.circuit.retained.rememberRetained
+
+/** Copy of [RetainedStateHolder] to return content value */
```

Comment:
nit: do you think we can/should reconcile these into one API?

Reply:
I think it would be good to consolidate them, but there are a few considerations:

1. Just like `SaveableStateHolder` in #1451 , we need to investigate further whether converting a `RetainedStateHolder` without a return value into one with a return value would cause any issues.

2. The `RetainedStateHolder` without a return value internally uses `ReusableContent(key)`, while the one with a return value uses `key(key)`. I’m not entirely sure if using `ReusableContent(key)` in the case of a return value would cause any issues.